KDOP_tree/include/CGAL/KDOP_tree/KDOP_kdop.h

```diff
@@ -0,0 +1,181 @@
+
+#include <vector>
+#include <algorithm>
+
+namespace CGAL {
+namespace KDOP_tree {
+
+/// \addtogroup PkgKDOPTree
+/// @{
+
+  /**
+   * Class KDOP_kdop is a data structure to compute and store k-dop of primitives.
+   */
+
+  template<typename GeomTraits>
+  class KDOP_kdop
+  {
+  public:
+    typedef GeomTraits R;
+
+    typedef typename R::Point_3 Point_3;
+    typedef typename R::Segment_3 Segment_3;
+    typedef typename R::Triangle_3 Triangle_3;
+
+    /// \name Types
+    /// @{
+
+    /// Type of directions
+    typedef std::vector<double> Direction_type;
+
+    /// Type of vector of k directions
+    typedef std::vector<Direction_type> Vec_direction;
+
+    /// Type of support heights
+    typedef std::vector<double> Vec_height;
+
+    typedef KDOP_kdop Kdop;
+
+    /// @}
+
+    /// \name Constructors
+    /// @{
+
+    /// Default constructor
+    KDOP_kdop() { }
+
+    /// Constructor with default directions
+    /// \todo Define default directions for some selected numbers N.
+    KDOP_kdop(unsigned int N) { }
+
+    /// Constructor with directions given.
+    KDOP_kdop(Vec_direction vector_direction)
+      : vector_direction_(vector_direction)
+    {}
+
+    /// @}
+
+    /// Inline function to set support heights in all directions.
+    void set_support_heights(const std::vector<double>& support_heights) { vector_height_ = support_heights; }
+
+    /// Inline function to return support heights in all directions.
+    Vec_height give_support_heights() const { return vector_height_; }
+
+    /// Function to compute support heights in all directions.
+    void compute_support_heights(const Triangle_3& t);
+
+    /// Inline function to return the minimum support height.
+    double min_height() const { return *std::min_element( vector_height_.begin(), vector_height_.end() ); }
+
+    /// Inline function to return the maximum support height.
+    double max_height() const { return *std::max_element( vector_height_.begin(), vector_height_.end() ); }
+
+    /// Inline function to set directions
+    void set_directions(const Vec_direction& vec_direction) { vector_direction_ = vec_direction; }
+
+    /// Inline function to return directions.
+    Vec_direction give_directions() const { return vector_direction_; }
+
+    /*!
+     * @brief Add a new direction to existing directions.
+     * @param new_direction the new direction to be added
+     * @return updated vector of (k + 1) directions
+     */
+    void add_direction(Direction_type new_direction) { vector_direction_.push_back(new_direction); }
+
+    /*!
+     * @brief Check if two k-dops overlap by comparing support heights of
+     * the two k-dops.
+     * @param kdop1 the first k-dop
+     * @param kdop2 the second k-dop
+     * @return true if the two k-dops overlap; otherwise, false.
+     * \todo Add the checking function.
+     */
+    bool do_overlap(const Kdop& kdop1, const Kdop& kdop2);
+
+  private:
+    Vec_direction vector_direction_;
```

Comment:
I do not understand why each kdop stores a vector with directions.  For the kdops in one tree they should all be the same.

Reply:
Yes, they should be the same. Maybe I can modify this so that the direction vector is imported as a parameter in member functions that require directions.

Reply:
Adapted the member functions so that it is not necessary to store direction vector in each kdop.